web-components/src/components/table-advanced/TableAdvanced.ts

```diff
@@ -0,0 +1,994 @@
+    }
+
+    // --------------
+
+    private onDropRow() {
+      if (this.dragRow != this.dropRow) {
+        this.drops.push([this.dragRow, this.dropRow]);
+      }
+
+      this.dragRow = -1;
+      this.dropRow = -1;
+      if (this.dragRowElem) {
+        this.dragRowElem.setAttribute("draggable", "false");
+        this.dragRowElem = null;
+      }
+    }
+
+    private onDropCol() {
+      if (this.dragCol != this.dropCol) {
+        const drag = this.COLS[this.dragCol];
+        const drop = this.COLS[this.dropCol];
+        const gDrag = drag.group;
+        const gDrop = drop.group;
+        const iDrag = drag.index;
+        const iDrop = drop.index;
+        drag.group = gDrop;
+        drop.group = gDrag;
+        drag.index = iDrop;
+        drop.index = iDrag;
+        this.COLS[this.dragCol] = drop;
+        this.COLS[this.dropCol] = drag;
+
+        this.ROWS.forEach(row => {
+          const rDrag = row[this.dragCol];
+          const rDrop = row[this.dropCol];
+          row[this.dragCol] = rDrop;
+          row[this.dropCol] = rDrag;
+        });
+
+        this.updCols();
+        this.updRows();
+      }
+
+      this.dragCol = -1;
+      this.dropCol = -1;
+    }
+
+    private sort(col: Col, order?: SortOrder) {
+      if (!col.sorter) return;
+
+      this.clearSelection();
+      this.drops = [];
+
+      this.COLS.forEach(c => {
+        if (c.options.id == col.options.id) {
+          c.sort = order ? order : c.sort == "default" ? "ascending" : c.sort == "ascending" ? "descending" : "default";
+        } else {
+          c.sort = "default";
+        }
+      });
+
+      this.updCols();
+      this.requestUpdate();
+
+      this["md-table-advanced-change"].emit({
+        type: "sort",
+        order: col.sort
+      });
+    }
+
+    @debounce(500)
+    private filter(col: Col) {
+      if (!col.filter) return;
+
+      this.clearSelection();
+      this.drops = [];
+
+      const input = col.filter.input.trim();
+      const filter = col.filter.list[col.filter.selectedIndex];
+
+      col.filter.active = input ? true : false;
+      this.requestUpdate();
+
+      if (input) {
+        this["md-table-advanced-change"].emit({
+          type: "filter-on",
+          filter,
+          input
+        });
+      } else {
+        this["md-table-advanced-change"].emit({
+          type: "filter-off",
+          filter: col.filter.list[col.filter.selectedIndex]
+        });
+      }
+    }
+
+    private selectRow(p: { row: Row; shiftKey: boolean; metaKey: boolean }) {
+      const i = p.row.idx;
+      const isSelected = this.selected.hasOwnProperty(i);
+      if (this.config.rows?.selectable == "multiple") {
+        if (p.metaKey) {
+          if (isSelected) {
+            delete this.selected[i];
+          } else {
+            this.selected[i] = true;
+          }
+        } else if (p.shiftKey) {
+          // TODO - if required
```

Comment:
empty use case?

Reply:
Oh i remember, can you add a longer comment describing the dilemma

Reply:
done